src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1408,46 +1413,105 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
-        private static int GetRank(Type type)
+        /// <summary>
+        /// Check that a parameter's rank and type are compatible with the given argument.
+        /// This function besides checking rank directly, also does checks using the FEP attributes.
+        /// </summary>
+        /// <param name="fep">the function end point wrapper</param>
+        /// <param name="paramType">type of param on the fep</param>
+        /// <param name="argType">actual argument type</param>
+        /// <param name="paramIndex">parameter index in list of params to fep.</param>
+        /// <returns></returns>
+        private static bool DoesParamRankMatchArg(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)
+        {
+            if (!paramType.Equals(argType))
+            {
+                //if both types are value types, let's
+                //let coercion figure this out.
+                if (paramType.IsValueType && argType.IsValueType)
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(fep,argType,paramIndex);
+                var paramRank = GetRank(fep,paramType, paramIndex);
+
+                //if the param is an arbitrary rank array and
+                //arg is some array type, replication should not occur.
+                if (paramRank == -1 && argRank > 0)
+                {
+                    return true;
+                }
+                if (argRank != paramRank) return false;
+
+                // If both have rank 0, it could also be because their type info is ambiguous.
+                //TODO commenting this out will makes negate test pass - as negate method expects object param.
+                //TODO perhaps just log something
+                //if (argRank == 0 && paramRank == 0) return false;
+
+                if (!paramType.IsAssignableFrom(argType))
+                {
+                    return false;
+                }
+
+            }
+            return true;
+        }
+
+        private static int GetRank(ProtoCore.CLRFunctionEndPoint fep,Type type,int paramIndex)
         {
+            //TODO (Alternatives to consider)
+            if (fep.procedureNode.ArgumentTypes.ElementAt(paramIndex).rank == DSASM.Constants.kArbitraryRank)
```

Comment:
while yes, that is true, IMO in the case of an arbitrary rank parameter, it does not matter what the arg rank is, we can mark them as compatible as we should not replicate.

If you prefer the alternative check that uses the individual arg/param types then I can switch to that, though I expect it will eventually get more complex or perhaps I can find a way of converting the clr type to a dstype and querying whether it is arbitrary rank.